### ISSUE TITLE
[rllib] high temporal difference error after restoring checkpoint - APEX-DQN/Torch

### ISSUE THREAD

**Issue by cedros23** [ray-project/ray#13132]
### What is the problem?
Hello, 

I have a custom env, a custom model setup that I train with Apex-DQN. I use tune.run and regularly save checkpoints (eg. 10 iterations). When I want to restore agent from a checkpoint by rollout, agent performs as expected with the high reward that matches with the checkpoint.

However, if I try to resume training, there is strange behavior. For the 1st iteration, the agent performs well again with a high reward average, but very high temporal difference error.

In the following iterations, reward drops quickly, and temporal difference error reduces too. It takes a reasonable amount of iterations to return the expected well-performing reward range again.

I tried different methods to restore agent/policy/checkpoint, which you can find below. In addition, I read many issues regarding restoring checkpoints and ran test_checkpoint_restore.py to confirm, my setup passes unit tests.
```

`# create a trainer:
trainer = dqn.ApexTrainer(config=config, env="motorwayenv:motorwayenv-v0")
trainer.restore(checkpoint_path=checkpoint_path)
model_weights = trainer.get_policy().get_weights()
policy = trainer.get_policy()
policy_state = policy.get_state()
print(policy_state)
# print(model_weights)
for i in range(0, 10):
    result = trainer.train()
    print(result)`
```
```
`# try
run("APEX", restore=checkpoint_path, config=config)
# another try
run("APEX", resume=True, local_dir=checkpoint_path)`
```

*Ray version and other system information (Python version, TensorFlow version, OS):*
ray version: 1.1.0.dev0 (build from source)

Edit I.
Upon further investigation, high TD error is due to different Target Network weights (below is the output from dqn_torch_policy.py):
It appears that the checkpoint correctly restores the weights of the Q network but not the target Q net. 

(pid=18034) Q network:  tensor([[6.1605, 6.4060, 6.5471],
(pid=18034)         [6.5344, 6.6905, 6.7796],
(pid=18034)         [6.5934, 6.4428, 4.9984],
(pid=18034)         [6.9184, 8.0263, 6.7440]])
(pid=18034) Q target:  tensor([[-0.0447, -0.0145,  0.0311],
(pid=18034)         [-0.0412, -0.0100,  0.0289],
(pid=18034)         [-0.0412, -0.0100,  0.0289],
(pid=18034)         [-0.0412, -0.0100,  0.0289]])

Edit II.

After additional tests, I can confirm that the target Q network is randomly initialized, although the actual network being restored from the checkpoint.. I will try to create a simple test script in cartpole env to check reproducibility.

**Comment by cedros23** [comment]
Hi again,

I believe that the issue would be reproduced with the following simple test script:

`if __name__ == "__main__":

    ray.init()
    experiment_analysis = tune.run(
        "APEX",
        stop={
            "training_iteration": 6,
        },
        config={
            "env": "CartPole-v0",
            "num_envs_per_worker": 1,
            "num_workers": 8,
            "framework": "torch",
            # Use GPUs iff `RLLIB_NUM_GPUS` env var set to > 0.
            "num_gpus": int(os.environ.get("RLLIB_NUM_GPUS", "0")),
        },
        checkpoint_freq=6)
    best_checkpoint = experiment_analysis.get_best_checkpoint(experiment_analysis.trials[0],
                                                              metric="episode_reward_mean",
                                                              mode="max")
    experiment_analysis = tune.run(
        "APEX",
        stop={
            "training_iteration": 12,
        },
        config={
            "env": "CartPole-v0",
            "num_envs_per_worker": 1,
            "num_workers": 8,
            "framework": "torch",
            # Use GPUs iff `RLLIB_NUM_GPUS` env var set to > 0.
            "num_gpus": int(os.environ.get("RLLIB_NUM_GPUS", "0")),
        },
        restore=best_checkpoint)

`
Due to random target network init during the restore, we have high temporal diff. error, which leads drastic reward drop right after restore:

![issue_1](https://user-images.githubusercontent.com/32811366/103464390-69054500-4d44-11eb-9903-045fc3d2a637.png)

![issue_2](https://user-images.githubusercontent.com/32811366/103464413-833f2300-4d44-11eb-97db-50961406dc4a.png)

**Comment by sven1977** [comment]
Thanks for filing this @cedros23 .
Here is a PR that fixes this problem. I was able to confirm your problem and tested the below PR. Weights for all Q-learning algos that have a `target_q_model` property will sync the target net after re-loading the weights form a checkpoint.
Let me know, if this doesn't work for you.

https://github.com/ray-project/ray/pull/14838

**Comment by sven1977** [comment]
Above PR has been merged.
